tests/unit/lms/views/admin/application_instance_test.py

```diff
@@ -2,28 +2,68 @@
 
 import pytest
 from pyramid.httpexceptions import HTTPNotFound
+from sqlalchemy.exc import IntegrityError
 
 from lms.services import ApplicationInstanceNotFound
-from lms.views.admin import AdminViews, logged_out, notfound
-from tests.matchers import temporary_redirect_to
+from lms.views.admin.application_instance import AdminApplicationInstanceViews
+from tests.matchers import Any, temporary_redirect_to
 
 
-@pytest.mark.usefixtures("pyramid_config", "application_instance_service")
-class TestAdminViews:
-    def test_index(self, pyramid_request, views):
-        response = views.index()
+@pytest.mark.usefixtures(
+    "pyramid_config", "application_instance_service", "lti_registration_service"
+)
+class TestAdminApplicationInstanceViews:
+    def test_instances(self, views):
+        assert views.instances() == {}
+
+    def test_instance_new(
+        self, pyramid_request_new_instance, application_instance_service
+    ):
+        response = AdminApplicationInstanceViews(
+            pyramid_request_new_instance
+        ).instance_new()
+
+        application_instance_service.create_application_instance.assert_called_once_with(
+            lms_url="LMS_URL",
+            email="EMAIL",
+            deployment_id="DEPLOYMENT_ID",
+            developer_key="",
+            developer_secret="",
+            lti_registration_id=sentinel.lti_registration_id,
+        )
 
         assert response == temporary_redirect_to(
-            pyramid_request.route_url("admin.instances")
+            pyramid_request_new_instance.route_url(
+                "admin.instance.id",
+                id_=application_instance_service.create_application_instance.return_value.id,
+            )
         )
 
-    def test_instances(self, views):
-        response = views.instances()
+    @pytest.mark.parametrize("missing", ["lms_url", "email", "deployment_id"])
+    def test_instance_new_missing_params(self, pyramid_request_new_instance, missing):
+        del pyramid_request_new_instance.params[missing]
+
+        response = AdminApplicationInstanceViews(
+            pyramid_request_new_instance
+        ).instance_new()
 
-        assert response == {}
+        assert response.status_code == 400
+
+    def test_instance_with_duplicate(
+        self, pyramid_request_new_instance, application_instance_service
+    ):
+        application_instance_service.create_application_instance.side_effect = (
+            IntegrityError(Any(), Any(), Any())
+        )
+
+        response = AdminApplicationInstanceViews(
+            pyramid_request_new_instance
+        ).instance_new()
+
+        assert response.status_code == 400
 
     def test_search_not_query(self, pyramid_request):
-        response = AdminViews(pyramid_request).search()
+        response = AdminApplicationInstanceViews(pyramid_request).search()
```

Comment:
There is a `views` fixture which looks like this. It looks like it would remove this repetition to use it.